lib/sla.js

```diff
@@ -34,62 +38,133 @@ function designateMixingBridge(client) {
-  addChannel({channel: channel.id})
+  addChannel({channel: [inbound.id, dialed.id]})
     .catch(function (err) {
       defer.reject(err);
     });
-  bridge.once('ChannelEnteredBridge', function(event, bridge) {
-    console.log('Channel %s has entered the bridge', bridge.channel.id);
-  });
-  bridge.once('ChannelLeftBridge', function(event, bridge) {
-    console.log('Channel %s has left the bridge', bridge.channel.id);
-    defer.resolve('Application completed');
-  });
+  function channelEntered(event, object) {
+    console.log('Channel %s has entered the bridge', object.channel.id);
+    numInBridge += 1;
+  }
+  bridge.on('ChannelEnteredBridge', channelEntered);
+  function channelLeft(event, object) {
+    numInBridge -= 1;
+    console.log('Channel %s has left the bridge', object.channel.id);
+    if(numInBridge === 1) {
+      if(object.channel.id === inbound.id) {
+        console.log('Hanging up dialed channel %s', dialed.id);
+        var hangupDialed = Q.denodeify(dialed.hangup.bind(dialed));
+        hangupDialed()
+          .catch(function (err) {
+            defer.reject(err);
+          });
+      }
+      else if(object.channel.id === dialed.id) {
+        console.log('Hanging up inbound channel %s', inbound.id);
+        var hangupInbound = Q.denodeify(inbound.hangup.bind(inbound));
+        hangupInbound()
+          .catch(function (err) {
+            defer.reject(err);
+          });
+      }
+    }
+    if(numInBridge === 0) {
+      defer.resolve('Application completed');
+      object.bridge.removeListener('ChannelEnteredBridge', channelEntered);
+      object.bridge.removeListener('ChannelLeftBridge', channelLeft);
+    }
+  };
+  bridge.on('ChannelLeftBridge', channelLeft);
   return defer.promise;
 }
 
 /**
- * Creates a channel to a specified endpoint and places it in the Stasis app.
+ * Originates a channel to a specified endpoint and places it in the Stasis app.
  * @param {Object} client - Object that contains information from the ARI
  *   connection.
+ * @param {Object} inbound- the inbound channel
  * @param {Object} bridge - Object that defines the bridge to be passed to
  *   addChannelsToBridge().
  * @return {Q} Q promise object.
  */
-function createChannel(client, bridge) {
+function originateChannel(client, inbound, bridge) {
   var defer = Q.defer();
-  var channel = client.Channel();
-  var originate = Q.denodeify(channel.originate.bind(channel));
-  originate({endpoint: 'SIP/phone', app: 'sla'})
+  var dialed = client.Channel();
+  var originate = Q.denodeify(dialed.originate.bind(dialed));
+  originate({endpoint: 'SIP/100@tcambron', app: 'sla', appArgs: 'dialed'})
     .catch(function (err) {
       defer.reject(err);
+      
     });
-  channel.once('StasisStart', function(event, channel) {
-    defer.resolve(channel);
+  dialed.once('StasisStart', function(event, dialed) {
+    var answer = Q.denodeify(dialed.answer.bind(dialed));
+    answer();
+    defer.resolve(dialed);
```

Comment:
You can't resolve the promise on both StasisStart and ChannelDestroyed. Once a promise has been resolved it's locked. There's no way for the caller to call .then for ChannelDestroyed.